### PR TITLE
[FEATURE] Add JointTorqueSensor.

### LINKED ISSUE #2983: [Feature]: Add joint torque sensor
### What feature or enhancement are you proposing?

Some robots have joint torque sensors placed in serie with their actuator. These sensor measure the total torque going through the joint: this can be viewed as being either:
 - the motor's input torque minus all actuator loss (friction, armature inertia, motor dynamics...)
 - the sum of all external forces acting on the joint: gravity, inertia and virtual forces, interaction forces...
 
Currently, this quantity is not readily available: Genesis gives access to
- `get_dofs_control_force`: motor torque output
- `get_dofs_force`: this looks like what I need, but when I compute it for a robot at rest, this quantity is close to zero - so it's at least missing the gravity term.

### Motivation

In order to implement a compliant controller on a collaborative robot (e.g. impedance control), some robot are built with torque sensors at the joint output. Feedback is then performed on this sensor to regulate the robot's torque. To simulate this controller, I need the same output available in simulation.

### Potential Benefit

Ability to simulate robots with joint torque sensors

### What is the expected outcome of the implementation work?

- [  ] A way to obtain the desired force- through a helper function  on a `RigidEntity` or a Sensor
- [  ] Unit test showing this is the quantity we want: maybe a robotic arm with and without interaction force to check both cases?

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by matthieuvigne**
## Description

Implement `JointTorqueSensor` class to measure the torque at the gearbox output. This is the motor's torque, minus losses in the gearbox: inertia, friction.

Resolves Genesis-Embodied-AI/Genesis#2983

## How Has This Been / Can This Be Tested?

Run the script  `examples/sensors/joint_torque_sensor_franka.py`, or the unit test file `tests/test_sensor_joint_torque.py`

### CODE REVIEW

**Comment by duburcqa** [comment]
```
0. Naming-convention fixes (flagged → corrected)

Files (redundant _sensor, directory inconsistency vs imu.py/contact_force.py/imu_franka.py):
- genesis/engine/sensors/joint_torque_sensor.py → joint_torque.py
- examples/sensors/joint_torque_sensor_franka.py → joint_torque_franka.py
- tests/test_sensor_joint_torque.py → deleted (folded into tests/test_sensors.py)

Classes:
- Options class JointTorqueSensor → JointTorque (options are nouns: IMU, Contact, ContactForce); engine class stays JointTorqueSensor.
- JointTorqueSensorData → removed (dead — single-tensor return never wraps it).
- JointTorqueSensorMetadata → kept (consistent with ContactForceSensorMetadata).

Kernel: _kernel_get_dofs_frictionloss_force (param fl_rank) → deleted entirely.

Locals (forbidden all_ prefix/suffix, fl abbreviations, vague/terse names, misleading aliases — all in code being rewritten): all_fl, rank_all, fl_rank, has_fl, qfrc_fl, dofs, n_total, qf_control → eliminated or replaced with conventional names; B = solver._B and entity._dof_start → public batch / dof_start API.

1. RigidSolver.get_dofs_actuator_force(dofs_idx=None, envs_idx=None) — new dev getter (core change)

Generalized effort at the actuator/gearbox output per DOF (torque for hinge, force for slide), kernel-free:
actuator_force = qf_applied - armature * qacc + qf_frictionloss + qf_passive
- Reads dofs_state.qf_applied, constraint_solver.qacc, dofs_state.qf_passive, dofs_info.armature via qd_to_torch(..., transpose=True) — no copy= (fresh arithmetic is returned; nothing internal aliases out).
- Frictionloss via approach (A), vectorized torch, no @qd.kernel: friction rows span [n_constraints_equality, n_constraints_equality + n_constraints_frictionloss), appended in ascending DOF-traversal order:
  - rank = cumsum(frictionloss > EPS, dim=-1) - 1
  - gather_idx = n_constraints_equality[:, None] + rank  ← document the friction-row-order == ascending-DOF-order invariant here
  - qf_frictionloss = gather(efc_force, gather_idx) * (frictionloss > EPS)
  - Reads only efc_force (B, len_con), n_constraints_equality (B,), frictionloss. Multi-env-correct (fixes the original (1,n) OOB in the right place).
- batch_dofs_info branch for armature/frictionloss (no envs_idx to the non-batched 1-D field); per-DOF terms read full then sliced once by dofs_idx; tensor[0] when n_envs == 0.
- Placed beside get_dofs_force / get_dofs_control_force; docstring distinguishes it (post-loss output) from get_dofs_control_force (commanded, pre-loss).

2. Sensor → thin wrapper (genesis/engine/sensors/joint_torque.py)

- Rename file; update engine/sensors/__init__.py import.
- _update_raw_data: actuator_force = solver.get_dofs_actuator_force(shared_metadata.dofs_idx); write actuator_force[None] if n_envs == 0 else actuator_force, transposed into raw_data_T. Delete the kernel, rank logic, get_dofs_frictionloss call.
- Remove the dead JointTorqueSensorData NamedTuple + NamedTuple import; return a bare tensor (as ContactForceSensor).
- build() uses public dof_start; metadata keeps solver + dofs_idx.
- ASCII-only docstrings/comments; drop the box-drawing separators.

3. Options (genesis/options/sensors/options.py)

- Add RigidEntitySensorOptionsMixin(SensorOptions[SensorT]): entity-level validation (entity_idx >= 0, in range, isinstance(RigidEntity)), no link fields — same diamond pattern as Contact(RigidSensorOptionsMixin, SimpleSensorOptions).
- Rename options class to JointTorque; JointTorque(RigidEntitySensorOptionsMixin[...], SimpleSensorOptions[...]) keeps dofs_idx_local + DOF-range / n_dofs == 0 checks via super().validate_scene().
- ASCII-only docstring; add the missing + tau_damping term; note "generalized effort (torque for hinge, force for slide)".

4. Tests — two functions folded into tests/test_sensors.py

- test_joint_torque — @pytest.mark.parametrize("n_envs", [0, 2]) × scenario {none, armature, frictionloss, damping}. Finite-diff loop; invariant actuator_force == tau_phys every scenario; scenario-gated extras (== tau_control, -I_arm*qddot, -damping*vel, the < inequalities).
- test_joint_torque_with_contact — @pytest.mark.parametrize("n_envs", [0, 2]). Pendulum + fixed wall; asserts actuator_force == tau_control (== tau_c) before/during/after impact (no finite-diff — impact makes qacc ill-defined).
- One shared module-level setup helper building the pendulum MJCF + scene from (armature, frictionloss, damping, wall) (2+ uses, non-trivial; _make_pendulum_xml inlined into it).
- No backend param, no docstrings (comments only), assert_allclose from tests/utils. Delete tests/test_sensor_joint_torque.py.

5. Example (examples/sensors/joint_torque_franka.py)

- Rename (drop _sensor); ASCII-only; fix the J0/J1/J2 vs "J0,J2,J4" label/comment mismatch; gs.logger instead of bare print; lazy matplotlib guard; drop the no-op -v flag.

6. Verification (worktree)

- First confirm the n_envs=2 failure reproduces on the original code, then run both tests on Metal + CPU; audit every touched call site before running.
```

**Review by duburcqa — DISMISSED** [review]

**Comment by duburcqa** [comment]
@codex review

File: genesis/options/sensors/options.py [hunk 1]
```diff
@@ -240,6 +253,40 @@ def model_post_init(self, context: Any) -> None:
...
+
+    Parameters
+    ----------
+    entity_idx : int
+        Scene-level index of the RigidEntity to sense. Must be >= 0.
+    dofs_idx_local : tuple[int, ...] | None, optional
+        Local DOF indices within the entity. ``None`` (default) selects all DOFs.
+    """
+
+    dofs_idx_local: tuple[int, ...] | None = None
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/options/sensors/options.py]: fixed
